src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -271,30 +284,112 @@ type Levels m h = V.Vector (Level m h)
 
 -- | Runs in order from newer to older
 data Level m h = Level {
-    incomingRuns :: !(MergingRun m h)
+    incomingRun  :: !(IncomingRun m h)
   , residentRuns :: !(V.Vector (Run m h))
   }
 
--- | A merging run is either a single run, or some ongoing merge.
-data MergingRun m h =
-    -- | A merging of multiple runs.
-    MergingRun
-      !MergePolicyForLevel
-      !NumRuns
-      !NumEntries
-      -- ^ Sum of number of entries in the input runs
-      !(UnspentCreditsVar (PrimState m))
-      -- ^ The number of currently /unspent/ credits
-      !(TotalStepsVar (PrimState m))
-      -- ^ The total number of performed merging steps.
-      !(MutVar (PrimState m) MergeKnownCompleted)
-      -- ^ A variable that caches knowledge about whether the merge has been
-      -- completed. If 'MergeKnownCompleted', then we are sure the merge has
-      -- been completed, otherwise if 'MergeMaybeCompleted' we have to check the
+-- | An incoming run is either a single run, or a merge.
+data IncomingRun m h =
+    Single !(Run m h)
+  | Merging !(MergingRun m h)
+
+-- | A merging of multiple runs.
+-- TODO: Move to a separate module.
+data MergingRun m h = MergingRun {
+      mergePolicy         :: !MergePolicyForLevel
+    , mergeNumRuns        :: !NumRuns
+      -- | Sum of number of entries in the input runs
+    , mergeNumEntries     :: !NumEntries
+      -- | The number of currently /unspent/ credits
+    , mergeUnspentCredits :: !(UnspentCreditsVar (PrimState m))
+      -- | The total number of performed merging steps.
+    , mergeStepsPerformed :: !(TotalStepsVar (PrimState m))
+      -- | A variable that caches knowledge about whether the merge has been
+      -- completed. If 'MergeKnownCompleted', then we are sure the merge has been
+      -- completed, otherwise if 'MergeMaybeCompleted' we have to check the
       -- 'MergingRunState'.
-      !(StrictMVar m (MergingRunState m h))
-    -- | The result of merging a single run, is a single run.
-  | SingleRun !(Run m h)
+    , mergeKnownCompleted :: !(MutVar (PrimState m) MergeKnownCompleted)
+    , mergeState          :: !(StrictMVar m (MergingRunState m h))
+    , mergeRefCounter     :: !(RC.RefCounter m)
```

Comment:
I think for `Merge` and (ultimately probably also `MergingRun`) the idea is to use a qualified import. It also contains identifiers like `new`.

Reply:
We can pick a different prefix, but I don't think it makes much of a difference. Also, IMO `mergingRun` is quite long and I don't particularly like abbreviations like `mr`.

Reply:
The advantage of using something like `mr` is that you can often create unique identifiers without sacrificing much of the identifier length. With qualified imports you might even choose to get rid of prefixes

I don't feel strongly about the names, I just wanted to bring up that `merge` is an overloaded name. It can refer to a `Merge` field, a `MergingRun` field, or a function (from the verb  "to merge")